receiver/k8sobjectsreceiver/receiver.go

```diff
@@ -148,36 +151,64 @@ func (kr *k8sobjectsreceiver) startPull(ctx context.Context, config *K8sObjectsC
 }
 
 func (kr *k8sobjectsreceiver) startWatch(ctx context.Context, config *K8sObjectsConfig, resource dynamic.ResourceInterface) {
-
 	stopperChan := make(chan struct{})
 	kr.mu.Lock()
 	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
 	kr.mu.Unlock()
 
-	resourceVersion, err := getResourceVersion(ctx, config, resource)
-	if err != nil {
-		kr.setting.Logger.Error("could not retrieve an initial resourceVersion", zap.String("resource", config.gvr.String()), zap.Error(err))
-		return
-	}
 	watchFunc := func(options metav1.ListOptions) (apiWatch.Interface, error) {
 		options.FieldSelector = config.FieldSelector
 		options.LabelSelector = config.LabelSelector
 		return resource.Watch(ctx, options)
 	}
 
+	cancelCtx, cancel := context.WithCancel(ctx)
+	cfgCopy := *config
+	wait.UntilWithContext(cancelCtx, func(newCtx context.Context) {
+		resourceVersion, err := getResourceVersion(newCtx, &cfgCopy, resource)
+		if err != nil {
+			kr.setting.Logger.Error("could not retrieve a resourceVersion", zap.String("resource", cfgCopy.gvr.String()), zap.Error(err))
+			cancel()
+			return
+		}
+
+		done := kr.doWatch(newCtx, &cfgCopy, resourceVersion, watchFunc, stopperChan)
+		if done {
+			cancel()
+			return
+		}
+
+		// need to restart with a fresh resource version
+		cfgCopy.ResourceVersion = ""
```

Comment:
Would it make sense to keep track of the last seen resource version so we don't miss any versions between when we received the 410 and when we did the list?

Reply:
Based on https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes I think a brand new list is the most robust solution, even if it means we might miss a couple pieces of data.  The RetryWatcher we use is keeping track of the latest ResourceVersion behind the scenes but chooses to give up anyways instead of restart with that value. Base on that current pattern I think it is safest to do a fresh list.